File: src/primitives/sensor_assertions.cpp
// Copyright 2026 John Vial
// SPDX-License-Identifier: Apache-2.0

#include "sim_harness/primitives/sensor_assertions.hpp"

#include <algorithm>
#include <cmath>
#include <sstream>

#include <sensor_msgs/point_cloud2_iterator.hpp>

namespace sim_harness::primitives
{

SensorDataResult assertSensorPublishing(
  rclcpp::Node::SharedPtr node,
  const std::string & topic,
  double expected_rate_hz,
  double tolerance_percent,
  std::chrono::seconds sample_duration)
{
  SensorDataResult result;
  result.valid = false;
  result.message_count = 0;
  result.publish_rate_hz = 0.0;

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);

  // Generic subscription that just counts messages
  size_t count = 0;
  auto sub = node->create_generic_subscription(
    topic, "std_msgs/msg/Header", 10,
    [&count](std::shared_ptr<rclcpp::SerializedMessage>) {
      ++count;
    });

  // Sample for the duration
  auto start = std::chrono::steady_clock::now();
  while (std::chrono::steady_clock::now() - start < sample_duration) {
    executor.spin_some(std::chrono::milliseconds(10));
  }

  result.message_count = count;
  double duration_sec = std::chrono::duration<double>(sample_duration).count();
  result.publish_rate_hz = static_cast<double>(count) / duration_sec;

  double deviation_percent = std::abs(result.publish_rate_hz - expected_rate_hz) /
    expected_rate_hz * 100.0;
  result.valid = deviation_percent <= tolerance_percent;

  std::stringstream ss;
  ss << "Rate: " << result.publish_rate_hz << " Hz";
  ss << " (expected: " << expected_rate_hz << " +/- " << tolerance_percent << "%)";
  ss << " - " << count << " messages in " << duration_sec << "s";
  result.details = ss.str();

  return result;
}

SensorDataResult assertLidarValid(
  rclcpp::Node::SharedPtr node,
  const std::string & topic,
  double min_range,
  double max_range,
  size_t min_points,
  std::chrono::seconds timeout)
{
  SensorDataResult result;
  result.valid = false;
  result.message_count = 0;

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);

  sensor_msgs::msg::PointCloud2::SharedPtr latest_msg;

  auto sub = node->create_subscription<sensor_msgs::msg::PointCloud2>(
    topic, 10,
    [&latest_msg, &result](const sensor_msgs::msg::PointCloud2::SharedPtr msg) {
      latest_msg = msg;
      result.message_count++;
    });

  // Wait for a message
  auto start = std::chrono::steady_clock::now();
  while (!latest_msg && std::chrono::steady_clock::now() - start < timeout) {
    executor.spin_some(std::chrono::milliseconds(10));
  }

  if (!latest_msg) {
    result.details = "No PointCloud2 messages received on " + topic;
    return result;
  }

  // Count valid points
  size_t total_points = latest_msg->width * latest_msg->height;
  size_t valid_points = 0;
  bool has_nan = false;

  try {
    sensor_msgs::PointCloud2ConstIterator<float> iter_x(*latest_msg, "x");
    sensor_msgs::PointCloud2ConstIterator<float> iter_y(*latest_msg, "y");
    sensor_msgs::PointCloud2ConstIterator<float> iter_z(*latest_msg, "z");

    for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z) {
      float x = *iter_x, y = *iter_y, z = *iter_z;

      if (std::isnan(x) || std::isnan(y) || std::isnan(z)) {
        has_nan = true;
        continue;
      }

      double range = std::sqrt(x * x + y * y + z * z);
      if (range >= min_range && range <= max_range) {
        valid_points++;
      }
    }
  } catch (const std::exception & e) {
    result.details = "Error reading point cloud: " + std::string(e.what());
    return result;
  }

  result.valid = valid_points >= min_points && !has_nan;

  std::stringstream ss;
  ss << "Points: " << valid_points << "/" << total_points << " valid";
  ss << " (range: " << min_range << "-" << max_range << "m)";
  if (has_nan) {
    ss << " [WARNING: NaN values detected]";
  }
  result.details = ss.str();

  return result;
}

SensorDataResult assertGpsValid(
  rclcpp::Node::SharedPtr node,
  const std::string & topic,
  double min_lat,
  double max_lat,
  double min_lon,
  double max_lon,
  std::chrono::seconds timeout)
{
  SensorDataResult result;
  result.valid = false;
  result.message_count = 0;

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);

  sensor_msgs::msg::NavSatFix::SharedPtr latest_msg;

  auto sub = node->create_subscription<sensor_msgs::msg::NavSatFix>(
    topic, 10,
    [&latest_msg, &result](const sensor_msgs::msg::NavSatFix::SharedPtr msg) {
      latest_msg = msg;
      result.message_count++;
    });

  auto start = std::chrono::steady_clock::now();
  while (!latest_msg && std::chrono::steady_clock::now() - start < timeout) {
    executor.spin_some(std::chrono::milliseconds(10));
  }

  if (!latest_msg) {
    result.details = "No NavSatFix messages received on " + topic;
    return result;
  }

  // Check for NaN
  bool has_nan = std::isnan(latest_msg->latitude) ||
    std::isnan(latest_msg->longitude) ||
    std::isnan(latest_msg->altitude);

  // Check bounds
  bool in_bounds =
    latest_msg->latitude >= min_lat && latest_msg->latitude <= max_lat &&
    latest_msg->longitude >= min_lon && latest_msg->longitude <= max_lon;

  // Check fix status
  bool has_fix = latest_msg->status.status >= sensor_msgs::msg::NavSatStatus::STATUS_FIX;

  result.valid = !has_nan && in_bounds && has_fix;

  std::stringstream ss;
  ss << "GPS: lat=" << latest_msg->latitude << ", lon=" << latest_msg->longitude;
  ss << ", alt=" << latest_msg->altitude;
  ss << ", status=" << static_cast<int>(latest_msg->status.status);
  if (has_nan) {ss << " [ERROR: NaN values]";}
  if (!in_bounds) {ss << " [ERROR: Out of bounds]";}
  if (!has_fix) {ss << " [WARNING: No fix]";}
  result.details = ss.str();

  return result;
}

SensorDataResult assertImuValid(
  rclcpp::Node::SharedPtr node,
  const std::string & topic,
  double max_acceleration,
  double max_angular_velocity,
  std::chrono::seconds timeout)
{
  SensorDataResult result;
  result.valid = false;
  result.message_count = 0;

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);

  sensor_msgs::msg::Imu::SharedPtr latest_msg;

  auto sub = node->create_subscription<sensor_msgs::msg::Imu>(
    topic, 10,
    [&latest_msg, &result](const sensor_msgs::msg::Imu::SharedPtr msg) {
      latest_msg = msg;
      result.message_count++;
    });

  auto start = std::chrono::steady_clock::now();
  while (!latest_msg && std::chrono::steady_clock::now() - start < timeout) {
    executor.spin_some(std::chrono::milliseconds(10));
  }

  if (!latest_msg) {
    result.details = "No IMU messages received on " + topic;
    return result;
  }

  // Check for NaN
  bool has_nan =
    std::isnan(latest_msg->linear_acceleration.x) ||
    std::isnan(latest_msg->linear_acceleration.y) ||
    std::isnan(latest_msg->linear_acceleration.z) ||
    std::isnan(latest_msg->angular_velocity.x) ||
    std::isnan(latest_msg->angular_velocity.y) ||
    std::isnan(latest_msg->angular_velocity.z);

  // Check acceleration magnitude
  double accel_mag = std::sqrt(
    std::pow(latest_msg->linear_acceleration.x, 2) +
    std::pow(latest_msg->linear_acceleration.y, 2) +
    std::pow(latest_msg->linear_acceleration.z, 2));

  // Check angular velocity magnitude
  double ang_vel_mag = std::sqrt(
    std::pow(latest_msg->angular_velocity.x, 2) +
    std::pow(latest_msg->angular_velocity.y, 2) +
    std::pow(latest_msg->angular_velocity.z, 2));

  bool accel_ok = accel_mag <= max_acceleration;
  bool ang_vel_ok = ang_vel_mag <= max_angular_velocity;

  result.valid = !has_nan && accel_ok && ang_vel_ok;

  std::stringstream ss;
  ss << "IMU: accel=" << accel_mag << " m/s^2, ang_vel=" << ang_vel_mag << " rad/s";
  if (has_nan) {ss << " [ERROR: NaN values]";}
  if (!accel_ok) {ss << " [ERROR: Acceleration too high]";}
  if (!ang_vel_ok) {ss << " [ERROR: Angular velocity too high]";}
  result.details = ss.str();

  return result;
}

SensorDataResult assertCameraValid(
  rclcpp::Node::SharedPtr node,
  const std::string & topic,
  uint32_t expected_width,
  uint32_t expected_height,
  const std::string & expected_encoding,
  std::chrono::seconds timeout)
{
  SensorDataResult result;
  result.valid = false;
  result.message_count = 0;

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);

  sensor_msgs::msg::Image::SharedPtr latest_msg;

  auto sub = node->create_subscription<sensor_msgs::msg::Image>(
    topic, 10,
    [&latest_msg, &result](const sensor_msgs::msg::Image::SharedPtr msg) {
      latest_msg = msg;
      result.message_count++;
    });

  auto start = std::chrono::steady_clock::now();
  while (!latest_msg && std::chrono::steady_clock::now() - start < timeout) {
    executor.spin_some(std::chrono::milliseconds(10));
  }

  if (!latest_msg) {
    result.details = "No Image messages received on " + topic;
    return result;
  }

  bool size_ok = true;
  if (expected_width > 0 && latest_msg->width != expected_width) {
    size_ok = false;
  }
  if (expected_height > 0 && latest_msg->height != expected_height) {
    size_ok = false;
  }

  bool encoding_ok = expected_encoding.empty() ||
    latest_msg->encoding == expected_encoding;

  bool data_ok = !latest_msg->data.empty();

  result.valid = size_ok && encoding_ok && data_ok;

  std::stringstream ss;
  ss << "Image: " << latest_msg->width << "x" << latest_msg->height;
  ss << ", encoding=" << latest_msg->encoding;
  ss << ", size=" << latest_msg->data.size() << " bytes";
  if (!size_ok) {ss << " [ERROR: Size mismatch]";}
  if (!encoding_ok) {ss << " [ERROR: Encoding mismatch]";}
  if (!data_ok) {ss << " [ERROR: Empty data]";}
  result.details = ss.str();

  return result;
}

SensorDataResult assertJointStatesValid(
  rclcpp::Node::SharedPtr node,
  const std::string & topic,
  const std::vector<std::string> & expected_joints,
  std::chrono::seconds timeout)
{
  SensorDataResult result;
  result.valid = false;
  result.message_count = 0;

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);

  sensor_msgs::msg::JointState::SharedPtr latest_msg;

  auto sub = node->create_subscription<sensor_msgs::msg::JointState>(
    topic, 10,
    [&latest_msg, &result](const sensor_msgs::msg::JointState::SharedPtr msg) {
      latest_msg = msg;
      result.message_count++;
    });

  auto start = std::chrono::steady_clock::now();
  while (!latest_msg && std::chrono::steady_clock::now() - start < timeout) {
    executor.spin_some(std::chrono::milliseconds(10));
  }

  if (!latest_msg) {
    result.details = "No JointState messages received on " + topic;
    return result;
  }

  // Check for expected joints
  std::vector<std::string> missing_joints;
  for (const auto & expected : expected_joints) {
    auto it = std::find(
      latest_msg->name.begin(), latest_msg->name.end(), expected);
    if (it == latest_msg->name.end()) {
      missing_joints.push_back(expected);
    }
  }

  // Check for NaN in values
  bool has_nan = false;
  for (double v : latest_msg->position) {
    if (std::isnan(v)) {has_nan = true; break;}
  }
  if (!has_nan) {
    for (double v : latest_msg->velocity) {
      if (std::isnan(v)) {has_nan = true; break;}
    }
  }
  if (!has_nan) {
    for (double v : latest_msg->effort) {
      if (std::isnan(v)) {has_nan = true; break;}
    }
  }

  result.valid = missing_joints.empty() && !has_nan;

  std::stringstream ss;
  ss << "JointStates: " << latest_msg->name.size() << " joints";
  if (!missing_joints.empty()) {
    ss << " [MISSING: ";
    for (size_t i = 0; i < missing_joints.size(); ++i) {
      ss << missing_joints[i];
      if (i < missing_joints.size() - 1) {ss << ", ";}
    }
    ss << "]";
  }
  if (has_nan) {ss << " [ERROR: NaN values]";}
  result.details = ss.str();

  return result;
}

}  // namespace sim_harness::primitives

File: sim_harness/primitives/sensor_assertions.py
# Copyright 2026 The sim_harness Authors
# SPDX-License-Identifier: Apache-2.0

"""
Sensor data validation assertions.

Provides functions to validate sensor data from ROS 2 topics.

Implementation uses :class:`TopicObserver` for subscription lifecycle
and composable predicates from :mod:`sim_harness.core.predicates` for
validation logic. All public signatures are unchanged.
"""

import math
from dataclasses import dataclass
from typing import List

from sensor_msgs.msg import NavSatFix, Imu, Image, JointState, LaserScan

from sim_harness.core.topic_observer import (
    collect_messages,
    count_messages,
    SENSOR_QOS,
)
from sim_harness.core.stream_properties import all_of
from sim_harness.core.predicates import (
    scan_has_min_points,
    scan_ranges_within,
    scan_nan_ratio_below,
    imu_no_nan,
    imu_accel_within,
    imu_gyro_within,
    image_has_data,
    image_dimensions,
    image_encoding,
    gps_no_nan,
    gps_in_bounds,
    gps_has_fix,
    joints_present,
    joints_no_nan,
)


@dataclass
class SensorDataResult:
    """Result of a sensor data validation."""

    valid: bool
    """Whether the sensor data is valid."""

    message_count: int
    """Number of messages received during sampling."""

    publish_rate_hz: float
    """Measured publish rate (Hz)."""

    details: str
    """Human-readable details."""


def assert_sensor_publishing(
    node,
    topic: str,
    expected_rate_hz: float,
    msg_type: type = LaserScan,
    tolerance_percent: float = 10.0,
    sample_duration_sec: float = 5.0,
) -> SensorDataResult:
    """
    Assert that a sensor is publishing at the expected rate.

    Args:
        node: ROS 2 node
        topic: Topic to monitor
        expected_rate_hz: Expected publish rate
        msg_type: Message type class for the topic (default: LaserScan)
        tolerance_percent: Acceptable deviation (percent)
        sample_duration_sec: How long to sample

    Returns:
        SensorDataResult with publish rate info
    """
    obs = count_messages(topic, msg_type, qos=SENSOR_QOS)
    result = obs.run_standalone(node, sample_duration_sec)

    message_count = result.value
    publish_rate = result.publish_rate_hz

    if expected_rate_hz > 0:
        deviation = abs(publish_rate - expected_rate_hz) / expected_rate_hz * 100
        valid = deviation <= tolerance_percent
    else:
        valid = message_count > 0
        deviation = 0

    details = (
        f"Rate: {publish_rate:.1f} Hz, Expected: {expected_rate_hz:.1f} Hz, "
        f"Deviation: {deviation:.1f}%"
    )

    return SensorDataResult(
        valid=valid,
        message_count=message_count,
        publish_rate_hz=publish_rate,
        details=details,
    )


def assert_lidar_valid(
    node,
    topic: str,
    min_range: float = 0.1,
    max_range: float = 100.0,
    min_points: int = 100,
    timeout_sec: float = 5.0,
) -> SensorDataResult:
    """
    Assert that LIDAR data is valid.

    Checks:
    - Messages are being received
    - Scan has minimum number of valid points
    - Range values are within bounds (no NaN, within sensor limits)

    Args:
        node: ROS 2 node
        topic: LIDAR topic (LaserScan message type)
        min_range: Minimum valid range (meters)
        max_range: Maximum valid range (meters)
        min_points: Minimum number of valid (non-NaN, non-inf) points expected
        timeout_sec: Time to wait for data

    Returns:
        SensorDataResult with validation status and point statistics
    """
    obs = collect_messages(topic, LaserScan, qos=SENSOR_QOS)
    result = obs.run_standalone(node, timeout_sec)
    messages = result.value

    if not messages:
        return SensorDataResult(
            valid=False,
            message_count=0,
            publish_rate_hz=0.0,
            details=f"No LIDAR messages received on {topic}",
        )

    scan = messages[-1]

    valid = all_of(
        scan_has_min_points(min_points),
        scan_ranges_within(min_range, max_range),
        scan_nan_ratio_below(0.1),
    )(scan)

    # Diagnostic details
    num_points = len(scan.ranges)
    nan_count = sum(1 for r in scan.ranges if math.isnan(r))
    inf_count = sum(1 for r in scan.ranges if math.isinf(r))
    valid_count = num_points - nan_count - inf_count
    out_of_range = sum(
        1 for r in scan.ranges
        if math.isfinite(r) and (r < min_range or r > max_range)
    )

    details = (
        f"Points: {num_points}, Valid: {valid_count}, "
        f"NaN: {nan_count}, Inf: {inf_count}, OutOfRange: {out_of_range}"
    )

    return SensorDataResult(
        valid=valid,
        message_count=result.message_count,
        publish_rate_hz=result.publish_rate_hz,
        details=details,
    )


def assert_gps_valid(
    node,
    topic: str,
    min_lat: float,
    max_lat: float,
    min_lon: float,
    max_lon: float,
    timeout_sec: float = 5.0,
) -> SensorDataResult:
    """
    Assert that GPS data is valid and within a region.

    Checks:
    - Messages are being received
    - Coordinates are within specified bounds
    - Fix status is valid
    - No NaN values in coordinates

    Args:
        node: ROS 2 node
        topic: GPS topic
        min_lat: Minimum latitude (degrees)
        max_lat: Maximum latitude (degrees)
        min_lon: Minimum longitude (degrees)
        max_lon: Maximum longitude (degrees)
        timeout_sec: Time to wait for data

    Returns:
        SensorDataResult
    """
    obs = collect_messages(topic, NavSatFix, qos=SENSOR_QOS)
    result = obs.run_standalone(node, timeout_sec)
    messages = result.value

    if not messages:
        return SensorDataResult(
            valid=False,
            message_count=0,
            publish_rate_hz=0.0,
            details=f"No GPS messages received on {topic}",
        )

    fix = messages[-1]

    valid = all_of(
        gps_no_nan(),
        gps_in_bounds(min_lat, max_lat, min_lon, max_lon),
        gps_has_fix(),
    )(fix)

    in_bounds = (
        min_lat <= fix.latitude <= max_lat
        and min_lon <= fix.longitude <= max_lon
    )

    details = (
        f"Lat: {fix.latitude:.6f}, Lon: {fix.longitude:.6f}, "
        f"Status: {fix.status.status}, InBounds: {in_bounds}"
    )

    return SensorDataResult(
        valid=valid,
        message_count=result.message_count,
        publish_rate_hz=result.publish_rate_hz,
        details=details,
    )


def assert_imu_valid(
    node,
    topic: str,
    max_acceleration: float = 50.0,
    max_angular_velocity: float = 10.0,
    timeout_sec: float = 5.0,
) -> SensorDataResult:
    """
    Assert that IMU data is valid.

    Checks:
    - Messages are being received
    - Acceleration values are reasonable (not extreme)
    - Angular velocity values are reasonable
    - No NaN values

    Args:
        node: ROS 2 node
        topic: IMU topic
        max_acceleration: Maximum expected acceleration (m/s^2)
        max_angular_velocity: Maximum expected angular velocity (rad/s)
        timeout_sec: Time to wait for data

    Returns:
        SensorDataResult
    """
    obs = collect_messages(topic, Imu, qos=SENSOR_QOS)
    result = obs.run_standalone(node, timeout_sec)
    messages = result.value

    if not messages:
        return SensorDataResult(
            valid=False,
            message_count=0,
            publish_rate_hz=0.0,
            details=f"No IMU messages received on {topic}",
        )

    imu = messages[-1]

    valid = all_of(
        imu_no_nan(),
        imu_accel_within(max_acceleration),
        imu_gyro_within(max_angular_velocity),
    )(imu)

    a = imu.linear_acceleration
    g = imu.angular_velocity
    accel_mag = math.sqrt(a.x**2 + a.y**2 + a.z**2)
    gyro_mag = math.sqrt(g.x**2 + g.y**2 + g.z**2)
    has_nan = not all(
        math.isfinite(v) for v in [a.x, a.y, a.z, g.x, g.y, g.z]
    )

    details = (
        f"Accel: {accel_mag:.2f} m/s^2, Gyro: {gyro_mag:.2f} rad/s, "
        f"NaN: {has_nan}"
    )

    return SensorDataResult(
        valid=valid,
        message_count=result.message_count,
        publish_rate_hz=result.publish_rate_hz,
        details=details,
    )


def assert_camera_valid(
    node,
    topic: str,
    expected_width: int = 0,
    expected_height: int = 0,
    expected_encoding: str = "",
    timeout_sec: float = 5.0,
) -> SensorDataResult:
    """
    Assert that camera images are valid.

    Checks:
    - Messages are being received
    - Image dimensions match expected (if specified)
    - Encoding matches (if specified)
    - Image data is non-empty

    Args:
        node: ROS 2 node
        topic: Camera image topic
        expected_width: Expected image width (0 = any)
        expected_height: Expected image height (0 = any)
        expected_encoding: Expected encoding (empty = any)
        timeout_sec: Time to wait for data

    Returns:
        SensorDataResult
    """
    obs = collect_messages(topic, Image, qos=SENSOR_QOS)
    result = obs.run_standalone(node, timeout_sec)
    messages = result.value

    if not messages:
        return SensorDataResult(
            valid=False,
            message_count=0,
            publish_rate_hz=0.0,
            details=f"No camera messages received on {topic}",
        )

    img = messages[-1]

    # Build predicate dynamically based on which checks are requested
    checks = [image_has_data()]
    if expected_width > 0 and expected_height > 0:
        checks.append(image_dimensions(expected_width, expected_height))
    if expected_encoding:
        checks.append(image_encoding(expected_encoding))

    valid = all_of(*checks)(img)

    details = (
        f"Size: {img.width}x{img.height}, Encoding: {img.encoding}, "
        f"Data: {len(img.data)} bytes"
    )

    return SensorDataResult(
        valid=valid,
        message_count=result.message_count,
        publish_rate_hz=result.publish_rate_hz,
        details=details,
    )


def assert_joint_states_valid(
    node,
    topic: str,
    expected_joints: List[str],
    timeout_sec: float = 5.0,
) -> SensorDataResult:
    """
    Assert that joint states contain expected joints.

    Checks:
    - Messages are being received
    - All expected joints are present
    - Position/velocity/effort values are not NaN

    Args:
        node: ROS 2 node
        topic: Joint states topic
        expected_joints: List of joint names that must be present
        timeout_sec: Time to wait for data

    Returns:
        SensorDataResult
    """
    obs = collect_messages(topic, JointState, qos=SENSOR_QOS)
    result = obs.run_standalone(node, timeout_sec)
    messages = result.value

    if not messages:
        return SensorDataResult(
            valid=False,
            message_count=0,
            publish_rate_hz=0.0,
            details=f"No joint state messages received on {topic}",
        )

    js = messages[-1]

    valid = all_of(
        joints_present(expected_joints),
        joints_no_nan(),
    )(js)

    missing_joints = [j for j in expected_joints if j not in js.name]

    details = (
        f"Joints: {js.name}, "
        f"Missing: {missing_joints if missing_joints else 'None'}, "
        f"NaN: {not joints_no_nan()(js)}"
    )

    return SensorDataResult(
        valid=valid,
        message_count=result.message_count,
        publish_rate_hz=result.publish_rate_hz,
        details=details,
    )
